shimmingtoolbox/shim/realtime_zshim.py

```diff
@@ -0,0 +1,259 @@
+    mean_p = np.mean(acq_pressures)
+    pressure_rms = np.sqrt(np.mean((acq_pressures - mean_p) ** 2))
+    riro = np.array([np.zeros_like(fieldmap[:, :, :, 0]),
+                     np.zeros_like(fieldmap[:, :, :, 0]),
+                     np.zeros_like(fieldmap[:, :, :, 0])])
+    static = np.array([np.zeros_like(fieldmap[:, :, :, 0]),
+                       np.zeros_like(fieldmap[:, :, :, 0]),
+                       np.zeros_like(fieldmap[:, :, :, 0])])
+    # TODO fix progress bar not showing up
+    progress_bar = st_progress_bar(fieldmap[..., 0].size * 3, desc="Fitting", ascii=False)
+    for g_axis in range(3):
+        for i_x in range(fieldmap.shape[0]):
+            for i_y in range(fieldmap.shape[1]):
+                for i_z in range(fieldmap.shape[2]):
+
+                    # do regression to separate static component and RIRO component
+                    reg = LinearRegression().fit(acq_pressures.reshape(-1, 1) - mean_p,
+                                                 -gradient[g_axis][i_x, i_y, i_z, :])
+                    # Multiplying by the RMS of the pressure allows to make abstraction of the tightness of the bellow
+                    # between scans. This allows to compare results between scans.
+                    riro[g_axis][i_x, i_y, i_z] = reg.coef_ * pressure_rms
+                    static[g_axis][i_x, i_y, i_z] = reg.intercept_
+                    progress_bar.update(1)
+
+    # Resample masked_fieldmaps to target anatomical image
+    nii_masked_fieldmaps = nib.Nifti1Image(masked_fieldmaps, nii_fieldmap.affine)
+    nii_resampled_fmap = resample_from_to(nii_masked_fieldmaps, nii_anat, order=2, mode='nearest')
+    if is_outputting_figures:
+        nib.save(nii_resampled_fmap, os.path.join(path_output, 'fig_resampled_fmap.nii.gz'))
+
+    # Resample static to target anatomical image
+    resampled_static = np.array([np.zeros_like(anat), np.zeros_like(anat), np.zeros_like(anat)])
+    for g_axis in range(3):
+        nii_static = nib.Nifti1Image(static[g_axis], nii_fieldmap.affine)
+        nii_resampled_static = resample_from_to(nii_static, nii_anat, mode='nearest')
+        resampled_static[g_axis] = nii_resampled_static.get_fdata()
+    # Since this is zshimming, only the slice component is used.
+    _, _, resampled_static_vox = phys_to_vox_gradient(resampled_static[0], resampled_static[1], resampled_static[2],
+                                                      nii_anat.affine)
+    nii_resampled_static_vox = nib.Nifti1Image(resampled_static_vox, nii_anat.affine)
+    nii_resampled_static_masked = nib.Nifti1Image(resampled_static_vox * nii_mask_anat.get_fdata(),
+                                                  nii_resampled_static_vox.affine)
+    if is_outputting_figures:
+        nib.save(nii_resampled_static_masked, os.path.join(path_output, 'fig_resampled_static.nii.gz'))
+
+    # Resample riro to target anatomical image
```

Comment:
high duplication with previous paragraph-- could potentially be a private local function